README.md
# Geo Location Service

> TYPO3 extension to resolve an address to a geo code via Google Maps API

This TYPO3 extension provides a geo coder service and an interface for geo codes.

## Installation

To install the Geo Location Service extension in your project, you have to require the package by using Composer:

```
composer require cpsit/geo-location-service
```

## Usage

You can implement the `GeoCodingInterface` in your location class and use the `GeoCoder` service to update or work with the geo information within that class. Check the source code for further information.


ChangeLog
2019-12-18         [RELEASE] Release of geo_location_service 0.3.0 (Dirk Wenzel)
2019-12-17 169de07 [RELEASE] Release of geo_location_service 0.3.0 (Elias Häußler)
2019-12-12 66a674e [BUGFIX] Encode URL parameters for geo location request correctly (Elias Häußler)
2019-12-10 d9a814f [BUGFIX] Ensure URI parameters are always encoded correctly (Elias Häußler)
2019-12-04 62b67eb [FEATURE] Allow passing custom parameters to service URL for API requests (Elias Häußler)
2019-12-03 b8c3b1d [TASK] Remove obsolete ext_localconf file (Elias Häußler)
2019-12-03 ce71fe1 [TASK] Fix version constraint for TYPO3 in ext_emconf.php (Elias Häußler)
2019-12-03 f9219e5 [TASK] Add platform requirements to composer.json (Elias Häußler)
2019-12-03 c10f56c [TASK] Use API to read extension configuration (Elias Häußler)
2019-11-14 4471a14 [FEATURE] Introduce GeoLocation cache to reduce API calls (Elias Häußler)
2019-10-15 f8adb54 Renamed 'Usage' to 'Installation' and wrote a short description about the actual usage of the extension. (j.fastnacht)

2019-12-17         [RELEASE] Release of geo_location_service 0.3.0 (Elias Häußler)
2019-12-12 66a674e [BUGFIX] Encode URL parameters for geo location request correctly (Elias Häußler)
2019-12-10 d9a814f [BUGFIX] Ensure URI parameters are always encoded correctly (Elias Häußler)
2019-12-04 62b67eb [FEATURE] Allow passing custom parameters to service URL for API requests (Elias Häußler)
2019-12-03 b8c3b1d [TASK] Remove obsolete ext_localconf file (Elias Häußler)
2019-12-03 ce71fe1 [TASK] Fix version constraint for TYPO3 in ext_emconf.php (Elias Häußler)
2019-12-03 f9219e5 [TASK] Add platform requirements to composer.json (Elias Häußler)
2019-12-03 c10f56c [TASK] Use API to read extension configuration (Elias Häußler)
2019-11-14 4471a14 [FEATURE] Introduce GeoLocation cache to reduce API calls (Elias Häußler)

2019-05-27         [RELEASE] Release of geo_location_service 0.2.0 (Dirk Wenzel)
2019-05-27 2574515 [BUGFIX] typo corrected (Dirk Wenzel)
2019-05-27 9ab07c2 [TASK] add ChangeLog (Dirk Wenzel)
2019-05-27 2d0fef8 [TASK] GeoCodingInterface renamed to GeoCodeableInterface (Dirk Wenzel)

2019-05-27         [RELEASE] Release of geo_location_service 0.1.0 (Dirk Wenzel)
2019-05-27 a51edfe [TASK] obsolete parameter removed rom serviceUrl (Dirk Wenzel)
2019-05-27 8b334b9 [TASK] extension manager configuration updated (Dirk Wenzel)
2019-05-27 a719880 [TASK] require typo3/cms-core:^9.5 (Dirk Wenzel)
2017-09-18 0c48482 [FEATURE] Add composer.json file (Nicole Cordes)
2017-08-09 f004252 [TASK] Geo service classes ported from t3events. ext_conf_template added. (Erik Rauchstein)


ext_localconf.php
<?php
defined('TYPO3_MODE') or die();

// Register geo location cache
(function() {
    $cacheConfigurations = &$GLOBALS['TYPO3_CONF_VARS']['SYS']['caching']['cacheConfigurations'];

    if (!is_array($cacheConfigurations[\CPSIT\GeoLocationService\Cache\GeoLocationCache::NAME])) {
        $cacheConfigurations[\CPSIT\GeoLocationService\Cache\GeoLocationCache::NAME] = [];
    }
    if (!isset($cacheConfigurations[\CPSIT\GeoLocationService\Cache\GeoLocationCache::NAME]['backend'])) {
        $cacheConfigurations[\CPSIT\GeoLocationService\Cache\GeoLocationCache::NAME]['backend']
            = \TYPO3\CMS\Core\Cache\Backend\Typo3DatabaseBackend::class;
    }
    if (!isset($cacheConfigurations[\CPSIT\GeoLocationService\Cache\GeoLocationCache::NAME]['options'])) {
        $cacheConfigurations[\CPSIT\GeoLocationService\Cache\GeoLocationCache::NAME]['options'] = [
            'defaultLifetime' => \CPSIT\GeoLocationService\Cache\GeoLocationCache::DEFAULT_LIFETIME,
        ];
    }
})();


Classes/Service/GeoCoder.php
<?php
namespace CPSIT\GeoLocationService\Service;

/***************************************************************
 *
 *  Copyright notice
 *
 *  (c) 2017 Erik Rauchstein <erik.rauchstein@cps-it.de>
 *  (c) 2019 Elias Häußler <e.haeussler@familie-redlich.de>
 *
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

use CPSIT\GeoLocationService\Cache\GeoLocationCache;
use CPSIT\GeoLocationService\Domain\Model\GeoCodableInterface;
use TYPO3\CMS\Core\Cache\Exception\NoSuchCacheException;
use TYPO3\CMS\Core\Configuration\ExtensionConfiguration;
use TYPO3\CMS\Core\Exception;
use TYPO3\CMS\Core\Http\Uri;
use TYPO3\CMS\Core\Utility\GeneralUtility;

/**
 * Geo coding service
 *
 * @license http://www.gnu.org/licenses/gpl.html GNU General Public License, version 3 or later
 */
class GeoCoder
{
    /**
     * @var array Valid URL parameters for Google Geocoding API
     * @see https://developers.google.com/maps/documentation/geocoding/intro#GeocodingRequests
     */
    public const VALID_SERVICE_URL_PARAMETERS = [
        'address',
        'key',
        'bounds',
        'language',
        'region',
        'components',
    ];

    /**
     * Service Url
     *
     * @var string Base Url for geo coding service.
     */
    protected $serviceUrl = 'https://maps.googleapis.com/maps/api/geocode/json?&address=';

    /**
     * Api key for geo code api
     *
     * @var string
     */
    protected $apiKey;

    /**
     * Configuration set by extension configuration.
     *
     * @var array
     */
    protected $extConf;

    /**
     * @var GeoLocationCache
     */
    protected $cache;

    /**
     * Returns the base url of the geo coding service
     *
     * @return string
     */
    public function getServiceUrl()
    {
        return $this->serviceUrl;
    }

    /**
     * Set the base url of the geo coding service
     *
     * @param $serviceUrl
     */
    public function setServiceUrl($serviceUrl)
    {
        $this->serviceUrl = $serviceUrl;
    }

    /**
     * @return string
     */
    public function getApiKey()
    {
        return $this->apiKey;
    }

    /**
     * @param string $apiKey
     */
    public function setApiKey($apiKey)
    {
        $this->apiKey = $apiKey;
    }

    public function __construct()
    {
        $this->cache = GeneralUtility::makeInstance(GeoLocationCache::class);

        try {
            $this->extConf = GeneralUtility::makeInstance(ExtensionConfiguration::class)->get('geo_location_service');
        } catch (Exception $e) {
            $this->extConf = [];
        }
        $this->setApiKey($this->extConf['googleApiKey']);
    }

    /**
     * Get geo location encoded from Google Maps geocode service.
     *
     * @param string $address An address to encode.
     * @param array $additionalParameters
     * @return array|false Array containing geo location information
     */
    public function getLocation($address, array $additionalParameters = [])
    {
        // Build request URI
        $apiParameters = array_merge($additionalParameters, [
            'address' => $address,
            'key' => $this->getApiKey(),
        ]);
        $url = $this->buildServiceUrlWithParameters($apiParameters);

        // Try to get geo location from cache
        $cacheIdentifier = $this->cache->calculateCacheIdentifier($url);
        try {
            $result = $this->cache->get($cacheIdentifier);
            if ($result !== false) {
                return $result;
            }
        } catch (NoSuchCacheException $e) {
            // Intended fallthrough if cache is not available.
        }

        $jsonResponse = $this->getUrl((string) $url);
        $response = json_decode($jsonResponse, true);

        if ($response['status'] !== 'OK') {
            return false;
        }

        $result = $response['results'][0]['geometry']['location'];

        try {
            $this->cache->set($cacheIdentifier, $result);
        } catch (NoSuchCacheException $e) {
            // Intended fallthrough if cache is not available.
        }

        return $result;
    }

    /**
     * @param array $parameters
     * @return Uri
     */
    public function buildServiceUrlWithParameters(array $parameters = []): Uri
    {
        $uri = new Uri($this->serviceUrl);

        if (empty($parameters)) {
            return $uri;
        }

        // Remove invalid URI parameters
        $parameters = array_filter($parameters, function ($parameterName) {
            return in_array($parameterName, self::VALID_SERVICE_URL_PARAMETERS);
        }, ARRAY_FILTER_USE_KEY);

        // Respect predefined parameters in service URI
        if (!empty($uri->getQuery())) {
            $parameters += GeneralUtility::explodeUrl2Array($uri->getQuery());
        }

        // Build URI with parameters
        $queryParams = urldecode(http_build_query($parameters));
        $uri = $uri->withQuery($queryParams);

        return $uri;
    }

    /**
     * Get url
     * Wrapper for GeneralUtility::getUrl to make it testable.
     *
     * @param string $url File/Url to fetch
     * @return mixed Response
     * @codeCoverageIgnore
     */
    public function getUrl($url)
    {
        return GeneralUtility::getUrl($url);
    }

    /**
     * calculate destination lat/lng given a starting point, bearing, and distance
     *
     * @param float $lat Latitude
     * @param float $lng Longitude
     * @param $bearing
     * @param integer $distance Distance
     * @param string $units Units: default km. Any other value will result in computing with mile based constants.
     * @return array An array with lat and lng values
     * @codeCoverageIgnore
     */
    public function destination($lat, $lng, $bearing, $distance, $units = 'km')
    {
        $radius = strcasecmp($units, 'km') ? 3963.19 : 6378.137;
        $rLat = deg2rad($lat);
        $rLon = deg2rad($lng);
        $rBearing = deg2rad($bearing);
        $rAngDist = $distance / $radius;

        $rLatB = asin(sin($rLat) * cos($rAngDist) +
            cos($rLat) * sin($rAngDist) * cos($rBearing));

        $rLonB = $rLon + atan2(sin($rBearing) * sin($rAngDist) * cos($rLat),
                cos($rAngDist) - sin($rLat) * sin($rLatB));

        return array('lat' => rad2deg($rLatB), 'lng' => rad2deg($rLonB));
    }

    /**
     * calculate bounding box
     *
     * @param float $lat Latitude of location
     * @param float $lng Longitude of location
     * @param float $distance Distance around location
     * @param string $units Unit: default km. Any other value will result in computing with mile based constants.
     * @return array An array describing a bounding box
     * @codeCoverageIgnore
     */
    public function getBoundsByRadius($lat, $lng, $distance, $units = 'km')
    {
        return array('N' => $this->destination($lat, $lng, 0, $distance, $units),
            'E' => $this->destination($lat, $lng, 90, $distance, $units),
            'S' => $this->destination($lat, $lng, 180, $distance, $units),
            'W' => $this->destination($lat, $lng, 270, $distance, $units));
    }

    /**
     * calculate distance between two lat/lon coordinates
     *
     * @param float $latA Latitude of location A
     * @param float $lonA Longitude of location A
     * @param float $latB Latitude of location B
     * @param float $lonB Longitude of location B
     * @param string $units Units: default km. Any other value will result in computing with mile based constants.
     * @return float
     * @codeCoverageIgnore
     */
    public function distance($latA, $lonA, $latB, $lonB, $units = 'km')
    {
        $radius = strcasecmp($units, 'km') ? 3963.19 : 6378.137;
        $rLatA = deg2rad($latA);
        $rLatB = deg2rad($latB);
        $rHalfDeltaLat = deg2rad(($latB - $latA) / 2);
        $rHalfDeltaLon = deg2rad(($lonB - $lonA) / 2);

        return 2 * $radius * asin(sqrt(pow(sin($rHalfDeltaLat), 2) +
            cos($rLatA) * cos($rLatB) * pow(sin($rHalfDeltaLon), 2)));
    }

    /**
     * Update Geo Location
     * Sets latitude and longitude of an object. The object
     * must implement the GeoCodableInterface.
     * Will first read city and zip attributes then tries to
     * get geo location values and if succeeds update the latitude and
     * longitude values of the object.
     *
     * @var GeoCodableInterface $object
     */
    public function updateGeoLocation(GeoCodableInterface &$object)
    {
        $city = $object->getPlace();
        if (!empty($city)) {
            $address = '';
            $zip = $object->getZip();
            $street = $object->getAddress();
            $address .= (!empty($zip)) ? $zip . ' ' : null;
            $address .= (!empty($street)) ? $street . ' ' : null;
            $address .= $city;
            $geoLocation = $this->getLocation($address);
            if ($geoLocation) {
                $object->setLatitude($geoLocation['lat']);
                $object->setLongitude($geoLocation['lng']);
            }
        }
    }
}


Classes/Cache/GeoLocationCache.php
<?php
declare(strict_types=1);
namespace CPSIT\GeoLocationService\Cache;

/*
 * This file is part of the TYPO3 CMS extension "geo_location_service".
 *
 * Copyright (C) 2019 Elias Häußler <e.haeussler@familie-redlich.de>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use TYPO3\CMS\Core\Cache\CacheManager;
use TYPO3\CMS\Core\Cache\Exception\NoSuchCacheException;
use TYPO3\CMS\Core\Cache\Frontend\FrontendInterface;
use TYPO3\CMS\Core\Http\Uri;
use TYPO3\CMS\Core\Utility\GeneralUtility;

/**
 * GeoLocationCache
 *
 * @author Elias Häußler <e.haeussler@familie-redlich.de>
 * @license GPL-2.0-or-later
 */
class GeoLocationCache
{
    /** @var string Cache name */
    public const NAME = 'geolocationservice_location';

    /** @var int Default cache lifetime */
    public const DEFAULT_LIFETIME = 86400;

    /**
     * @var FrontendInterface Cache instance
     */
    private $cache;

    /**
     * Initialize GeoLocation cache.
     *
     * @throws NoSuchCacheException
     */
    public function __construct()
    {
        $this->cache = GeneralUtility::makeInstance(CacheManager::class)->getCache(self::NAME);
    }

    /**
     * @param string $cacheIdentifier
     * @return mixed|null
     */
    public function get(string $cacheIdentifier)
    {
        return $this->cache->get($cacheIdentifier);
    }

    /**
     * @param string $cacheIdentifier
     * @param $data
     * @param array $tags
     * @param int|null $lifetime
     */
    public function set(string $cacheIdentifier, $data, array $tags = [], int $lifetime = null): void
    {
        $this->cache->set($cacheIdentifier, $data, $tags, $lifetime);
    }

    /**
     * @param Uri $serviceUrl
     * @return string
     */
    public function calculateCacheIdentifier(Uri $serviceUrl): string
    {
        $queryParams = GeneralUtility::explodeUrl2Array($serviceUrl->getQuery());
        array_multisort($queryParams);
        return sha1((string) $serviceUrl->withQuery(http_build_query($queryParams)));
    }
}


Classes/Domain/Model/GeoCodingInterface.php
<?php
namespace CPSIT\GeoLocationService\Domain\Model;

/***************************************************************
 *
 *  Copyright notice
 *
 *  (c) 2017 Erik Rauchstein <erik.rauchstein@cps-it.de>
 *
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

interface GeoCodingInterface
{
    public function getPlace();

    public function getZip();

    public function getAddress();

    public function getLatitude();

    public function getLongitude();

    public function setLatitude($latitude);

    public function setLongitude($longitude);
}


Classes/Domain/Model/GeoCodableInterface.php
<?php
namespace CPSIT\GeoLocationService\Domain\Model;

/***************************************************************
 *
 *  Copyright notice
 *
 *  (c) 2017 Erik Rauchstein <erik.rauchstein@cps-it.de>
 *
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

interface GeoCodableInterface
{
    public function getPlace();

    public function getZip();

    public function getAddress();

    public function getLatitude();

    public function getLongitude();

    /**
     * @param $latitude
     * @return void
     */
    public function setLatitude($latitude);

    /**
     * @param $longitude
     * @return mixed
     */
    public function setLongitude($longitude);
}


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "geo_location_service".
 *
 * Auto generated 27-09-2016 12:04
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = [
    'title' => 'Geo Location Service',
    'description' => 'Get latitude and longitude for an address per google api call',
    'category' => 'be',
    'author' => 'Erik Rauchstein',
    'author_email' => 'erik.rauchstein@cps-it.de',
    'state' => 'beta',
    'uploadfolder' => 0,
    'createDirs' => '',
    'clearCacheOnLoad' => 0,
    'author_company' => '',
    'version' => '0.3.0',
    'constraints' => [
        'depends' => [
            'typo3' => '9.5.0-9.5.99',
        ],
        'suggests' => [
            'pagenotfoundhandling' => '',
        ],
    ],
];

